runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -719,3 +750,710 @@ TR::TreeLowering::lowerArrayStoreCHK(TR::Node *node, TR::TreeTop *tt)
+      {
+      TR::Node *deps = helperCallBlock->getExit()->getNode()->getChild(0);
+      helperCallBlock->getExit()->getNode()->setNumChildren(0);
+      deps->decReferenceCount();
+      gotoAfterHelperCallNode->addChildren(&deps, 1);
+      }
+   }
+
+/*
+ * lowerStoreArrayElement transforms the block that contains the jitStoreFlattenableArrayElement helper call into three blocks:
+ *   1. The merge block that contains the tree tops after the helper call
+ *   2. The helper call block that contains the helper call and is moved to the end of the tree top list
+ *   3. The new non-VT array store block which is an extended block of the original block
+ *
+ *      originalBlock ----------+
+ *      arrayElementStoreBlock  |
+ *               |              |
+ *               |              v
+ *               |       helperCallBlock
+ *               |              |
+ *               v              v
+ *             blockAfterHelperCall
+ *
+ Before:
+ +-------------------------------------------+
+ |treetop                                    |
+ |   acall  jitStoreFlattenableArrayElement  |
+ |      aload <value>                        |
+ |      iload <index>                        |
+ |      aload <arrayAddress>                 |
+ |ttAfterHelperCall                          |
+ +-------------------------------------------+
+
+ After:
+ +-------------------------------------------+
+ |BBStart                                    |
+ |treetop                                    |
+ |   aload <ArrayAddress>                    |
+ |treetop                                    |
+ |   aload <index>                           |
+ |treetop                                    |
+ |   aload <value>                           |
+ |ificmpne ------>---------------------------+---------------+
+ |   iand                                    |               |
+ |      iloadi  <isClassFlags>               |               |
+ |      ...                                  |               |
+ |      iconst 1024                          |               |
+ |   iconst 0                                |               |
+ |   GlRegDeps ()                            |               |
+ |      PassThrough rcx                      |               |
+ |         ==>aload                          |               |
+ |      PassThrough r8                       |               |
+ |         ==>aload                          |               |
+ |      PassThrough rdi                      |               |
+ |         ==>iload                          |               |
+ |BBEnd                                      |               |
+ +-------------------------------------------+               |
+ +-------------------------------------------+               |
+ |BBStart (extension of previous block)      |               |
+ |NULLCHK on n82n                            |               |
+ |   ...                                     |               |
```

Comment:
I think this `NULLCHK` will only be generated if the store call was anchored in a `NULLCHK`.  Do have that right?  At the risk of cluttering things, you might just want to show something like this, just to make it clear that it won't always be generated.

```
NULLCHK on n82n  [if required]
```

Reply:
>I think this NULLCHK will only be generated if the store call was anchored in a NULLCHK. Do have that right?

Correct